specification/dartLangSpec.tex

```diff
@@ -5204,28 +5204,42 @@ \subsubsection{Correct Member Overrides}
   $F$ must then be a subtype of $F'$.
 
   \commentary{%
-  The subtype requirement ensures that argument list shapes
-  that are admissible for an invocation of a method with signature $m'$
-  are also admissible for an invocation of a method with signature $m$.
-  For instance, $m'$ may accept 2 or 3 positional arguments,
-  and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
-  This is a built-in property of the function type subtype rules.
-  %
-  Note that a member signature differs from
-  an underlying syntactic declaration $D$ in a class $C$.
-  In particular,
-  a parameter in a member signature has the modifier \COVARIANT{}
-  if and only if the parameter is covariant-by-declaration
-  (\ref{covariantParameters}),
-  and that may be the case due to declarations in a supertype of $C$,
-  so that modifier need not be present in $D$.
-  %
-  There is an additional potential compile-time error associated with
-  a parameter which is covariant-by-declaration
-  (\ref{instanceMethods}).
-  But we cannot cover that here as a property of member overrides,
-  because it is concerned with declarations in all superinterfaces,
-  indirect as well as direct.%
+    The subtype requirement ensures that argument list shapes
+    that are admissible for an invocation of a method with signature $m'$
+    are also admissible for an invocation of a method with signature $m$.
+    For instance, $m'$ may accept 2 or 3 positional arguments,
+    and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
+    This is a built-in property of the function type subtype rules.
+
+    Note that a member signature differs from
+    an underlying syntactic declaration $D$ in a class $C$.
+    In particular,
+    a parameter in a member signature has the modifier \COVARIANT{}
+    if and only if the parameter is covariant-by-declaration
+    (\ref{covariantParameters}),
+    and that may be the case due to declarations in a supertype of $C$,
+    so that modifier need not be present syntactically in $D$.
+  }
+
+  It is a compile-time error if $m'$ has a parameter
+  with the modifier \COVARIANT,
```

Comment:
We are in a predicate definition ("is a valid override if .... "), so it's the wrong place to introduce compile-time errors. Just say that m is not a correct override of m'.
It's then up to other places in the spec to make it an error if something is not a valid override of something else.

*Also*, I think that if m has a top type as parameter type, it is a valid override of a covariant function. After all, it can accept all the same arguments.

Reply:
> Just say ..
Good point! Done.

> .. top type ..
Good point, too!

This is likely to be a new behavior for the analyzer (as well as the CFE), so this means that there will be an implementation effort in the analyzer that wasn't previously expected for this spec change. However, there's no rush to do it because it is a non-breaking change. A developer who _would_ have used this feature would always be able to write a forwarding method (which is what they need to do in the non-top type case, anyway).